File: src/goserver/nf/game/food.go
/**
 * Created by nazarigonzalez on 17/10/16.
 */

package game

import (
	"../utils"
	"gopkg.in/nazariglez/go-easing.v0"
	"gopkg.in/nazariglez/go-rbush.v0"
	"math"
	"math/rand"
)

const (
	rad55                     = 55 * utils.DEG_TO_RAD
	foodOutTime               = 5.5
	foodMinRadius             = 31
	foodMaxRadius             = 80
	foodMaxMassRelativeRadius = 410
)

var (
	foodRadiusEasing = easing.OutExpo()
)

type Food struct {
	State EntityState

	Id                         int
	Mass, Angle, Radius, Speed float64

	Box      *rbush.Box
	Position *Point

	Data *FoodData

	OnExpireOut func(f *Food)

	vectorSpeed            *Point
	decreaseSpeed, outTime float64
	out                    bool
}

func NewEntityFood() *Food {
	box := rbush.Box{}

	return &Food{
		Id:       -1,
		State:    STATE_ALIVE,
		Box:      &box,
		Position: &Point{},
		Data:     &FoodData{},

		vectorSpeed: &Point{},

		outTime: foodOutTime,
	}
}

func (f *Food) Initialize(id int, mass float64) {
	f.Id = id
	f.Box.Data = id

	if mass == 0 {
		f.SetMass(1)
	} else {
		f.SetMass(mass)
	}
}

func (f *Food) Reset() {
	f.Id = -1
	f.SetMass(0)
	f.Box.MinX = 0
	f.Box.MinY = 0
	f.Box.MaxX = 0
	f.Box.MaxY = 0
	f.Box.Data = -1
	f.State = STATE_ALIVE
	f.Angle = 0
	f.Speed = 0
	f.Position.Set(0, 0)
	f.vectorSpeed.Set(0, 0)
	f.out = false
	f.decreaseSpeed = 0
	f.outTime = foodOutTime
}

func (f *Food) SetMass(m float64) {
	if m != f.Mass {
		f.Mass = m
		f.Radius = CalculateFoodRadius(f.Mass)
	}
}

func (f *Food) SetSpeed(v float64) {
	if v != f.Speed {
		cosX := math.Cos(f.Angle)
		sinY := math.Sin(f.Angle)

		f.vectorSpeed.Set(
			v*cosX,
			v*sinY,
		)

		f.Speed = v
	}
}

func (f *Food) SetMovement(angle, speed float64) {
	min := -rad55 + angle

	f.State = STATE_BORN
	f.Angle = min + (rand.Float64() * (rad55 * 2))
	f.SetSpeed(speed + easing.Interpolate(0, 50, foodMaxMassRelativeRadius, f.Mass, easing.Linear()))
	f.decreaseSpeed = speed * 0.8
}

func (f *Food) Update(delta float64) {
	if f.State == STATE_BORN {
		if f.Speed <= 20 {
			f.State = STATE_ALIVE
		}
	}

	if f.Speed > 0 {
		f.Position.X += f.vectorSpeed.X * delta
		f.Position.Y += f.vectorSpeed.Y * delta

		f.SetSpeed(f.Speed - f.decreaseSpeed*delta)
	}

	if f.out {
		f.outTime -= delta
		if f.outTime <= 0 && f.State == STATE_ALIVE {
			f.OnExpireOut(f)
		}
	}
}

func (f *Food) FixPosition(delta, x, y float64) {
	if f.out {
		return
	}

	if f.Position.X < -x || f.Position.X > x || f.Position.Y < -y || f.Position.Y > y {
		f.out = true
	}
}

func (f *Food) UpdateBox() {
	f.Box.MinX = f.Position.X - f.Radius
	f.Box.MinY = f.Position.Y - f.Radius
	f.Box.MaxX = f.Position.X + f.Radius
	f.Box.MaxY = f.Position.Y + f.Radius
}

func (f *Food) ToData() *FoodData {
	f.Data.Id = f.Id
	f.Data.Type = ENTITY_TYPE_FOOD
	f.Data.X = math.Floor(f.Position.X)
	f.Data.Y = math.Floor(f.Position.Y)
	f.Data.Radius = math.Floor(f.Radius)
	f.Data.Mass = math.Floor(f.Mass)
	return f.Data
}

func CalculateFoodRadius(m float64) float64 {
	if m == 1 {
		return 22
	} else if m > foodMaxMassRelativeRadius {
		return foodMaxRadius
	}

	return easing.Interpolate(foodMinRadius, foodMaxRadius, foodMaxMassRelativeRadius, m, foodRadiusEasing)
}

type FoodData struct {
	Id                 int
	Type               EntityType
	X, Y, Radius, Mass float64
}

func (d *FoodData) GetCreateBuffer() []byte {
	x := int(d.X)
	y := int(d.Y)
	r := int(d.Radius)
	m := int(d.Mass)

	return []byte{
		byte(d.Id >> 8 & 0xff),
		byte(d.Id >> 0 & 0xff),

		byte(ENTITY_TYPE_FOOD),

		byte(m >> 8 & 0xff),
		byte(m >> 0 & 0xff),

		byte(x >> 8 & 0xff),
		byte(x >> 0 & 0xff),

		byte(y >> 8 & 0xff),
		byte(y >> 0 & 0xff),

		byte(r >> 8 & 0xff),
		byte(r >> 0 & 0xff),
	}
}

func (d *FoodData) GetUpdateBuffer() []byte {
	x := int(d.X)
	y := int(d.Y)

	return []byte{
		byte(d.Id >> 8 & 0xff),
		byte(d.Id >> 0 & 0xff),

		byte(x >> 8 & 0xff),
		byte(x >> 0 & 0xff),

		byte(y >> 8 & 0xff),
		byte(y >> 0 & 0xff),
	}
}

func (d *FoodData) Copy() FoodData {
	return FoodData{
		Id:     d.Id,
		Type:   d.Type,
		X:      d.X,
		Y:      d.Y,
		Radius: d.Radius,
		Mass:   d.Mass,
	}
}

func (d *FoodData) ShouldUpdate(d2 *FoodData) bool {
	return d.X != d2.X || d.Y != d2.Y
}

File: src/server/Food.ts
/**
 * Created by nazarigonzalez on 28/6/16.
 */
import Point from './Point';
import {FOOD_DATA, ENTITY_TYPE} from "../common/interfaces";
import * as rbush from "rbush";
import ObjPool from "obj-pool";
import {interpolate, Easing} from "./interpolator";
import {arrayPool} from "../common/utils";

enum STATE {
  BORN, ALIVE, DEAD
}

export enum TYPE {
  FOOD, FIRE
}

const rad55:number = 55*Math.PI/180;
const OUT_TIME:number = 5.5;
const MIN_RADIUS:number = 31;
const MAX_RADIUS:number = 80;
const MAX_MASS_RELATIVE_RADIUS:number = 410;

export class Food {
  static pool: ObjPool<any> = new ObjPool(Food);
  static STATE = STATE;
  static TYPE = TYPE;

  static calculateRadius(mass: number) {
    if(mass === 1)return 22;
    return mass > MAX_MASS_RELATIVE_RADIUS ? MAX_RADIUS : interpolate(MIN_RADIUS, MAX_RADIUS, MAX_MASS_RELATIVE_RADIUS, mass, Easing.outExpo());
  };

  private _state: STATE = STATE.ALIVE;
  protected _type: TYPE = TYPE.FOOD;
  private _vectorSpeed: Point = new Point();
  protected _angle: number = 0;
  protected _speed: number = 0;
  private _decreaseSpeed: number = 0;
  private _outTime: number = OUT_TIME;
  private _out: boolean = false;
  private _mass: number = 0;
  private _radius: number = 0;

  id: number;
  position: Point = new Point();
  box: rbush.ItemInterface = {minX: 0, minY: 0, maxX: 0, maxY: 0, id: -1};

  constructor() {
  }

  initialize(id: number, mass?: number) {
    if (!mass) {
      this.mass = 1;// + Math.floor(Math.random() * 2);
    } else {
      this.mass = mass;
    }

    this.id = id;
    this.box["id"] = this.id;
  }

  reset() {
    this.id = -1;
    this.mass = 0;
    this.box.minX = 0;
    this.box.minY = 0;
    this.box.maxX = 0;
    this.box.maxY = 0;
    this.box["id"] = -1;
    this.state = Food.STATE.ALIVE;
    this._angle = 0;
    this._speed = 0;
    this.position.set(0);
    this._vectorSpeed.set(0);
    this._out = false;
    this._decreaseSpeed = 0;
    this._outTime = OUT_TIME;
  }

  onExpireOut(food: Food) {
  }

  setMovement(angle, speed) {
    this._state = STATE.BORN;
    const min: number = -rad55 + angle;
    this._angle = min + (Math.random() * (rad55 * 2));
    this.speed = speed + interpolate(0, 200, MAX_MASS_RELATIVE_RADIUS, this._mass);
    this._decreaseSpeed = speed * 0.8;
  }

  fixPosition(delta: number, xBound: number, yBound: number) {
    if (this._out)return;
    if (this.position.x < -xBound || this.position.x > xBound || this.position.y < -yBound || this.position.y > yBound) {
      this._out = true;
    }
  }

  update(delta) {
    if (this._state === STATE.BORN) {
      if(this.speed <= 20){
        this._state = STATE.ALIVE;
      }
    }

    if (this.speed > 0) {
      this.position.x += this._vectorSpeed.x * delta;
      this.position.y += this._vectorSpeed.y * delta;

      this.speed -= this._decreaseSpeed * delta;
    }

    if (this._out) {
      this._outTime -= delta;
      if (this._outTime <= 0 && this._state === STATE.ALIVE) {
        this.onExpireOut(this);
      }
    }
  }

  updateBox() {
    this.box.minX = this.position.x - this.radius;
    this.box.minY = this.position.y - this.radius;
    this.box.maxX = this.position.x + this.radius;
    this.box.maxY = this.position.y + this.radius;
  }

  toArray(): FOOD_DATA {
    let arr: any[] = arrayPool.alloc();
    arr[0] = this.id;
    arr[1] = ENTITY_TYPE.FOOD;
    arr[2] = Math.round(this.mass);
    arr[3] = Math.round(this.position.x);
    arr[4] = Math.round(this.position.y);
    arr[5] = Math.round(this.radius);
    return arr as FOOD_DATA;
  }

  get state(): STATE {
    return this._state;
  };

  set state(value: STATE) {
    if (value !== this._state) {
      this._state = value;
    }
  }

  get speed(): number {
    return this._speed;
  };

  set speed(value: number) {
    if (value !== this._speed) {
      const cosX: number = Math.cos(this._angle);
      const sinY: number = Math.sin(this._angle);
      this._vectorSpeed.set(
        value * cosX,
        value * sinY
      );
      this._speed = value;
    }
  }

  get type(): TYPE {
    return this._type;
  };

  get mass(): number {
    return this._mass;
  };

  set mass(value: number) {
    if (value !== this._mass) {
      this._mass = value;
      this._radius = Food.calculateRadius(this._mass);
    }
  }

  get radius(): number {
    return this._radius;
  };
}